torch/testing/_internal/opinfo/definitions/linalg.py

```diff
@@ -0,0 +1,1180 @@
+    make_fullrank = make_fullrank_matrices_with_distinct_singular_values
+    make_arg = partial(
+        make_fullrank, dtype=dtype, device=device, requires_grad=requires_grad
+    )
+
+    is_linalg_svd = "linalg.svd" in op_info.name
+    batches = [(), (0,), (3,)]
+    ns = [0, 3, 5]
+
+    def uniformize(usv):
+        S = usv[1]
+        k = S.shape[-1]
+        U = usv[0][..., :k]
+        Vh = usv[2] if is_linalg_svd else usv[2].mH
+        Vh = Vh[..., :k, :]
+        return U, S, Vh
+
+    def fn_U(usv):
+        U, _, _ = uniformize(usv)
+        return U.abs()
+
+    def fn_S(usv):
+        return uniformize(usv)[1]
+
+    def fn_Vh(usv):
+        # We also return S to test
+        _, S, Vh = uniformize(usv)
+        return S, Vh.abs()
+
+    def fn_UVh(usv):
+        U, S, Vh = uniformize(usv)
+        return U @ Vh, S
+
+    fns = (fn_U, fn_S, fn_Vh, fn_UVh)
+
+    fullmat = "full_matrices" if is_linalg_svd else "some"
+
+    for batch, n, k, fullmat_val, fn in product(batches, ns, ns, (True, False), fns):
+        shape = batch + (n, k)
+        yield SampleInput(
+            make_arg(*shape), kwargs={fullmat: fullmat_val}, output_process_fn_grad=fn
+        )
+
+
+def sample_inputs_cross(op_info, device, dtype, requires_grad, **kwargs):
+    yield SampleInput(
+        make_tensor((S, 3), device=device, dtype=dtype, requires_grad=requires_grad),
+        args=(
+            make_tensor(
+                (S, 3), device=device, dtype=dtype, requires_grad=requires_grad
+            ),
+        ),
+    )
+    yield SampleInput(
+        make_tensor((S, 3, S), device=device, dtype=dtype, requires_grad=requires_grad),
+        args=(
+            make_tensor(
+                (S, 3, S), device=device, dtype=dtype, requires_grad=requires_grad
+            ),
+        ),
+        kwargs={"dim": 1},
+    )
+    yield SampleInput(
+        make_tensor((S, 3), device=device, dtype=dtype, requires_grad=requires_grad),
+        args=(
+            make_tensor(
+                (S, 3), device=device, dtype=dtype, requires_grad=requires_grad
+            ),
+        ),
+        kwargs={"dim": -1},
+    )
+
+
+def sample_inputs_householder_product(op_info, device, dtype, requires_grad, **kwargs):
+    """
+    This function generates input for torch.linalg.householder_product (torch.orgqr).
+    The first argument should be a square matrix or batch of square matrices, the second argument is a vector or batch of vectors.
+    Empty, square, rectangular, batched square and batched rectangular input is generated.
+    """
+    # Each column of the matrix is getting multiplied many times leading to very large values for
+    # the Jacobian matrix entries and making the finite-difference result of grad check less accurate.
+    # That's why gradcheck with the default range [-9, 9] fails and [-2, 2] is used here.
+    samples = (
+        SampleInput(
+            make_tensor(
```

Comment:
that formatting takes a lot of space...

Reply:
Very true. IIRC the reasoning is that adding a new element in the middle of a list shouldn't require the entire list to be re-wrapped and therefore show up as changed in the diff. I've definitely seen lots of cases where merge conflicts could have been avoided if we used this style uniformly. So I think it's a reasonable trade-off.

Reply:
Ok!